---
title: "Marker Effect x Environment Analysis"
output: html_notebook
---

## Introduction

This notebook will look at "painting" the chromosomes to determine regions with large contributions to the mean and those with large contributions to GxE.

Load packages and data

```{r setup}

library(tidyverse)
library(stringr)
library(rrBLUP)
library(FW)
library(readxl)
library(neyhart)
library(gws)
library(gridExtra)
library(cowplot)


# The head directory
proj_dir <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET/"

# Prediction directory
pred_dir <- file.path(proj_dir, "Predictions")

# Other directories
geno_dir <-  "C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomic Selection/Genotypic Data/GBS Genotype Data/"
env_var_dir <- file.path(proj_dir, "Environmental_Variables/")
pheno_dir <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET/Phenotype_Data/"


## Load Data
# Load the genotypic data
load(file.path(geno_dir, "S2_genos_mat.RData"))

# Load the phenotypic data
load(file.path(pheno_dir, "S2_MET_BLUE.RData"))
load(file.path(pheno_dir, "S2_MET_tidy.RData"))

# Load environmental data
load(file.path(env_var_dir, "environmental_data_compiled.RData"))

# Load an entry file
entry_file <- file.path(proj_dir, "Plant_Materials/S2MET_project_entries.xlsx")
entry_list <- read_excel(entry_file)

# Get rid of checks
entries <- entry_list %>% 
  filter(Class != "Check") %>%
  pull(Line)

# Pull out the tp lines that intersect with the genotype data
tp <- entries %>% 
  str_subset("^[0-9]{2}") %>%
  intersect(., row.names(s2_imputed_mat))

vp <- entries %>% 
  str_subset("^2MS") %>%
  intersect(., row.names(s2_imputed_mat))

# Keep only the relevant lines
# Other phenotype data manipulations

## We need to combine some environments and prepare others for prediction, including:
# The environments "CNY15" and "HNY15" are the same

pheno <- S2_MET_BLUE %>%
  filter(line_name %in% c(tp, vp)) %>%
  mutate(environment = ifelse(environment == "CNY15", "HNY15", as.character(environment))) %>%
  # Change line_name and environment to factors
  mutate_at(vars(line_name, environment), funs(as.factor)) %>%
  droplevels()


```



Calculate mean genotypic effects and FW stability coefficients using the following model:

$$
y_{ij} = \mu + g_i + b_ih_j + \epsilon_{ij}
$$

where $y_{ij}$ is the phenotypic observation of the *i*th genotype in the *j*th environment, $\mu$ is the grand mean, $g_i$ is the main effect of the *i*th genotype, $b_i$ is the sensitivity of the *i*th genotype to the effect of the *j*th environment ($h_j$), and $\epsilon_{ij}$ is the residual.




```{r}

# Grab genotypes for the training population
geno_train <- s2_imputed_mat[tp,]

# Grab phenotypes for the training population
pheno_train <- pheno %>%
  filter(line_name %in% tp) %>%
  droplevels()

# Perform the regression
pheno_train_mod <- pheno_train %>% 
  group_by(trait) %>% 
  do(fw = FW(y = .$value, VAR = .$line_name, ENV = .$environment, method = "OLS"))

# Extract genotypic main effects and sensitivity measures
pheno_train_effs <- pheno_train_mod %>% 
  group_by(trait) %>% 
  do({
    .$fw[[1]][c("g", "b")] %>% 
      reduce(cbind) %>% 
      data.frame(line_name = row.names(.), ., row.names = NULL, stringsAsFactors = FALSE) %>% 
      rename(g = X1, b = X2)
  })

# For each trait, calculate marker effects
marker_eff <- pheno_train_effs %>% 
  gather(coef, value, -trait, -line_name) %>%
  group_by(trait, coef) %>%
  do({
    
    # Mixed model
    y <- as.matrix(.$value)
    X <- model.matrix(~ 1, data = .)
    
    # Solve
    out <- mixed.solve(y = y, Z = geno_train)
    out$u %>% 
      data.frame(rs = row.names(.), effect = ., row.names = NULL, stringsAsFactors = FALSE)

    })

# Add snp_info
marker_eff_ann <- full_join(rename(snp_info, rs = `rs#`), marker_eff) %>%
  mutate(chrom = as.factor(chrom))

# Group by trait and coef and create a matrix of marker effects per genotype
mar_eff_df <- marker_eff_ann %>% 
  group_by(trait, coef) %>%
  do({
    tp_ga = .$effect * t(geno_train)
    tp_ga %>% 
      data.frame(rs = row.names(.), ., row.names =  NULL, 
                 stringsAsFactors = FALSE, check.names = FALSE) %>% 
      gather(line_name, effect, -rs) %>%
      tbl_df() })

# Join again with the marker information
mar_eff_df_ann <- full_join(rename(snp_info, rs = `rs#`), mar_eff_df) %>%
  mutate(chrom = as.factor(chrom))


```



Summarize by chromosome. Sum the marker effects carried by each individual over
the whole chromosome

```{r ga.chrom}

mar_eff_chrom_summ <- mar_eff_df_ann %>%
  group_by(trait, coef, chrom, line_name) %>% 
  summarize(mar_eff_sum = sum(effect)) %>%
  # Add breeding program information
  left_join(., subset(entry_list, select = c(Line, Program)), by = c("line_name" = "Line"))
  

# Create a plot per trait and coefficient
mar_eff_plot <- mar_eff_chrom_summ %>% 
  group_by(trait, coef) %>% 
  do(plot = {
    ggplot(data = ., aes(x = line_name, y = chrom, fill = mar_eff_sum)) +
      geom_tile() +
      scale_fill_gradient2(low = "blue", high = "red",
                           guide = guide_colorbar(title = NULL)) +
      # space = "free_x" keeps the x-axis width equal
      facet_grid(trait + coef ~Program, scales = "free_x", space = "free_x") +
      ylab("") +
      xlab("Genotype") + 
      theme(
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        strip.text.x = element_blank(),
        plot.margin=unit(c(-0.1,1,-0.1,1),"lines")
      ) })

# Adjust the first plot
mar_eff_plot$plot[[1]] <- mar_eff_plot$plot[[1]] +
  theme(
    strip.text.x = element_text(size = 10)
  )

plot_grid(plotlist = mar_eff_plot$plot, align = "hv", ncol = 1)
      


```




```{r}





## Create blocks of n cM
cM_window <- 5

## Sum the marker effects in each block
marker_eff_summ <- marker_eff_ann %>% 
  group_by(chrom, trait) %>% 
  do({
    # Chromosome breaks
    breaks <- seq(0, max(.$cM_pos) + cM_window, by = cM_window)
    
    split(x = ., cut(.$cM_pos, breaks = seq(0, max(.$cM_pos) + cM_window, by = cM_window))) %>%
      map(~summarize(., n_SNP = n(), g_effect_sum = sum(g_effect), b_effect_sum = sum(b_effect))) %>% 
      bind_rows() %>% 
      mutate(cM_break = head(breaks, -1)) }) %>%
  gather(effect_type, effect, -chrom:-n_SNP, -cM_break)

# Plot the chromosomes
g <- marker_eff_summ %>% 
  group_by(chrom, trait, effect_type) %>% 
  summarize(max_break = max(cM_break) + cM_window) %>%
  ggplot(aes(x = chrom, y = max_break, group = effect_type, fill = effect_type)) +
  geom_col(fill = c("grey70"), position = "dodge") +
  facet_grid(.~ trait)


marker_eff_summ %>%
  ggplot(aes(fill = g_effect_sum), group = effect_type) + 
  geom_rect(aes(xmin = chrom - 0.45, xmax = chrom + 0.45, ymin = cM_break, 
                ymax = cM_break + cM_window)) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red")


marker_eff_summ %>%
  filter(trait == "PlantHeight") %>%
  ggplot(aes(x = chrom, y = abs(effect))) +
  geom_point() +
  facet_grid(effect_type ~ chrom, scales = "free_y")



```


